runtime/vm/createramclass.cpp

```diff
@@ -1529,17 +1529,26 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 				return FALSE;
 			}
 
+			UDATA superclassModifiers = superclass->romClass->modifiers;
 			/* ensure that the superclass isn't an interface or final */
-			if ((superclass->romClass->modifiers & J9_JAVA_FINAL) != 0) {
+			if (J9_ARE_ALL_BITS_SET(superclassModifiers, J9AccFinal)) {
 				Trc_VM_CreateRAMClassFromROMClass_superclassIsFinal(vmThread, superclass);
 				setCurrentExceptionForBadClass(vmThread, superclassName, J9VMCONSTANTPOOL_JAVALANGVERIFYERROR);
 				return FALSE;
 			}
-			if ((superclass->romClass->modifiers & J9_JAVA_INTERFACE) != 0) {
+			if (J9_ARE_ALL_BITS_SET(superclassModifiers, J9AccInterface)) {
 				Trc_VM_CreateRAMClassFromROMClass_superclassIsInterface(vmThread, superclass);
 				setCurrentExceptionForBadClass(vmThread, superclassName, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR);
 				return FALSE;
 			}
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			/* ensure that the superclass isn't a value type */
+			if (J9_ARE_ALL_BITS_SET(superclassModifiers, J9AccValueType)) {
```

Comment:
the super class needs to be java.lang.Object

Reply:
That is a separate check, correct? This checks that a class doesn't inherit from a value type class.

Also, why do all of these checks throw an error with the superclass named as the bad class? Isn't the class inheriting from the interface/value type the malformed one?

Reply:
> That is a separate check, correct? This checks that no class inherits from a value type class.
ValueTypes can't inherit anything other than j.l.Object. The reasoning is that they that they can't have polymorphic layouts. So they only inherit from the base class.

> Also, why do all of these checks throw an error with the superclass named as the bad class?
We assume the currentClass was compiled properly and the error is with the dependencies

Reply:
>ValueTypes can't inherit anything other than j.l.Object. The reasoning is that they that they can't have polymorphic layouts. So they only inherit from the base class.

I added this check below but the previous check is doing something different. From the spec:
>If the class or interface named as the direct superclass of C is in fact
an interface or a value class, loading throws
an IncompatibleClassChangeError.

(where C is a nonarray class or interface)

Reply:
youre right, didnt notice that.